core/modules/filter/filter.module

```diff
@@ -2520,20 +2520,27 @@ function _filter_image_align($text) {
       else {
         $target = $node;
       }
-      $classes = $target->getAttribute('class');
-      $classes = (strlen($classes) > 0) ? explode(' ', $classes) : array();
-      $classes[] = 'align-' . $align;
-      $target->setAttribute('class', implode(' ', $classes));
 
       // Center aligned content requires an additional wrapper to set
       // text-align: center.
       if ($align === 'center') {
-        $wrapper_div = $dom->createElement('div');
-        $wrapper_attribute = $dom->createAttribute('class');
-        $wrapper_attribute->value = 'centered-wrapper';
-        $wrapper_div->appendChild($wrapper_attribute);
-        $wrapper_div->appendChild($target->cloneNode(TRUE));
-        $target->parentNode->replaceChild($wrapper_div, $target);
+
+        // If the image does not have a caption get the closest p parent.
+        if ($parent->nodeName != 'figure') {
+          while ($target->nodeName !== 'p') {
+            $target = $target->parentNode;
+          }
+        }
+
+        $parent_attribute = $dom->createAttribute('class');
+        $parent_attribute->value = 'centered-wrapper';
+        $target->appendChild($parent_attribute);
```

Comment:
Will this blow away any other classes that are on the closest P element? It seems like it should do the same get/setAttribute() approach as the `else` statement below.